### PR TITLE
C++: Fix join in `isUse`

### PR DESCRIPTION
**Pull request by MathiasVP**
Before:
```ql
[2023-07-05 11:58:43] Evaluated non-recursive predicate UseAfterFree#36ea366b::isUse#2#ff@046a25uf in 2850ms (size: 952783).
Evaluated relational algebra for predicate UseAfterFree#36ea366b::isUse#2#ff@046a25uf with tuple counts:
    8043415   ~3%    {2} r1 = SCAN DataFlowUtil#47741e1f::OperandNode#fff OUTPUT In.0, In.2
    8043415   ~0%    {3} r2 = JOIN r1 WITH DataFlowUtil#47741e1f::OperandNode#fff ON FIRST 1 OUTPUT Rhs.2, Lhs.0, Lhs.1
    1428436   ~0%    {4} r3 = JOIN r2 WITH Instruction#577b6a83::CallInstruction::getArgumentOperand#fff_201#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Rhs.2
    1428058   ~0%    {4} r4 = JOIN r3 WITH Instruction#577b6a83::CallInstruction::getStaticCallTarget#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3
  108038889   ~0%    {4} r5 = JOIN r4 WITH Instruction#577b6a83::Instruction::getEnclosingFunction#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3
      70213   ~6%    {4} r6 = JOIN r5 WITH UseAfterFree#36ea366b::ParameterSinks::getAnAlwaysDereferencedParameter#0#f ON FIRST 1 OUTPUT Lhs.0, Lhs.3, Lhs.1, Lhs.2
      32865   ~4%    {2} r7 = JOIN r6 WITH Instruction#577b6a83::InitializeParameterInstruction::hasIndex#ff ON FIRST 2 OUTPUT Lhs.3, Lhs.2
      32864   ~0%    {2} r8 = JOIN r7 WITH Operand#27b358ab::Operand::getDef#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1
      32864   ~0%    {2} r9 = JOIN r8 WITH Instruction#577b6a83::Instruction::getUnconvertedResultExpression#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1
                  
      952783   ~0%    {2} r10 = UseAfterF
```

After:
```ql
Evaluated relational algebra for predicate UseAfterFree#36ea366b::IsUse::isUse#2#ff@8e45a80r with tuple counts:
    10601   ~0%    {2} r1 = JOIN DataFlowUtil#47741e1f::InstructionNode#fff_20#join_rhs WITH UseAfterFree#36ea366b::ParameterSinks::getAnAlwaysDereferencedParameter#0#f ON FIRST 1 OUTPUT Lhs.0, Lhs.1
    10601   ~3%    {2} r2 = JOIN r1 WITH Instruction#577b6a83::Instruction::getEnclosingFunction#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1
    32890   ~0%    {3} r3 = JOIN r2 WITH DataFlowImplCommon#59e7a193::Cached::viableParamArg#3#fff_102#join_rhs ON FIRST 1 OUTPUT Rhs.2, Lhs.1, Rhs.1
    32890   ~1%    {4} r4 = JOIN r3 WITH DataFlowUtil#47741e1f::OperandNode#fff ON FIRST 1 OUTPUT Lhs.2, Lhs.1, Lhs.0, Rhs.2
    32865   ~6%    {2} r5 = JOIN r4 WITH Instruction#577b6a83::CallInstruction::getStaticCallTarget#0#dispred#ff ON FIRST 2 OUTPUT Lhs.3, Lhs.2
    32864   ~0%    {2} r6 = JOIN r5 WITH Operand#27b358ab::Operand::getDef#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1
    32864   ~0%    {2} r7 = JOIN r6 WITH Instruction#577b6a83::Instruction::getUnconvertedResultExpression#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1
              
  952783   ~0%    {2} r8 = UseAfterFree#36ea366b::isUse0#2#ff UNION r7
                  return r8
```

### CODE REVIEW

File: cpp/ql/src/Critical/UseAfterFree.ql [hunk 1]
```diff
@@ -135,18 +135,24 @@ module ParameterSinks {
...
-  or
-  exists(CallInstruction call, int i, InitializeParameterInstruction init |
-    n.asOperand().getDef().getUnconvertedResultExpression() = e and
-    init = ParameterSinks::getAnAlwaysDereferencedParameter() and
-    call.getArgumentOperand(i) = n.asOperand() and
-    init.hasIndex(i) and
-    init.getEnclosingFunction() = call.getStaticCallTarget()
-  )
+module IsUse {
+  private import semmle.code.cpp.ir.dataflow.internal.DataFlowImplCommon
```
Comment (jketema) [inline cpp/ql/src/Critical/UseAfterFree.ql:139]: I don't particularly like that we have to pull in something internal. Is this really necessary?

Comment (jketema) [reply cpp/ql/src/Critical/UseAfterFree.ql:139]: Let's not re-invent anything, but if do wonder if it's really necessary to pull this in. Why don't the introduced pragmas suffice?

Comment (MathiasVP) [reply cpp/ql/src/Critical/UseAfterFree.ql:139]: We discussed this in a call and decided to have this internal import here for now with the goal of:
1. Identifying other places where we're doing this parameter-argument mapping manually or using this internal import, and
2. Fixing this with a non-internal predicate.

Comment (jketema) [reply cpp/ql/src/Critical/UseAfterFree.ql:139]: We briefly discussed this one-on-one, and pulling in the internal predicate seems the best solution at the moment without duplicating a lot of code from that internal predicate. So I'll accept this as-is.

As the original non-optimal join-order might also occur in other queries that have similar logic, we'll open an internal issue to track this and to provide a proper predicate that is not internal and efficient.

**Review by jketema — APPROVED** [review]

**Pull request merged**